### PR TITLE
Group conversations

### PR DESCRIPTION
**Pull request by BRHoward**
- Added functionality for group conversations
- Once a conversation has started users can add other users and leave conversations
- Users can rename conversations
- Added tests
- Project could do with refactoring/splitting up but didn't want pull request to become any bigger, this will be the focus of the next pull request

### CODE REVIEW

**Comment by BRHoward** [comment]
@stevenhillcox @jebbinSCL @sievins

File: public/index.html [hunk 1]
```diff
@@ -3,14 +3,19 @@
     <head>
         <meta charset="utf-8">
         <title>Chat</title>
-        <script src="https://ajax.googleapis.com/ajax/libs/angularjs/1.5.8/angular.min.js"></script>
-        <script src="//ajax.googleapis.com/ajax/libs/angularjs/1.5.8/angular-animate.js"></script>
+        <script src="http://ajax.googleapis.com/ajax/libs/angularjs/1.5.8/angular.min.js"></script>
+        <script src="http://ajax.googleapis.com/ajax/libs/angularjs/1.5.8/angular-animate.min.js"></script>
```
Comment (jebbinSCL) [inline public/index.html:8]: I think you will want to look into some form of bundler now.

File: public/js/MainController.js [hunk 2]
```diff
@@ -1,19 +1,19 @@
-/*global console, _*/
+    /* global _, app */
 
-(function () {
-    var app = angular.module("ChatApp", ["ngAnimate", "toastr"]);
-
-    app.controller("ChatController", function ($scope, $http, toastr) {
+    app.controller("ChatController", function ($scope, $http, $mdDialog, toastr) {
```
Comment (jebbinSCL) [inline public/js/MainController.js]: Is this file only the ChatController? If it isn't the chat controller should be split out into a ChatController.js file. Or this needs to be renamed to ChatController.js

File: public/index.html [hunk 3]
```diff
@@ -3,19 +3,18 @@
     <head>
         <meta charset="utf-8">
         <title>Chat</title>
+        <!-- external libraries !-->
         <script src="http://ajax.googleapis.com/ajax/libs/angularjs/1.5.8/angular.min.js"></script>
```
Comment (jebbinSCL) [inline public/index.html:7]: In a separate pull request, you may as well bundle your libraries as well! You can either do this as a separate bundle, or just have a single one for the entire app.

Comment (BRHoward) [reply public/index.html:7]: Okay, would that be purely to improve readability of the HTML page? Would there be any other advantages considering the external libraries are already optimized/minified?

Comment (jebbinSCL) [reply public/index.html:7]: Ah, actually you probably don't want to do that for another reason. Bundling them would stop them from being cached nicely as well. You would possibly do that in another case, but you can ignore it here.

**Comment by jebbinSCL** [comment]
:ship:
You'll want to start splitting up / refactoring your test code as well, since the file is getting big.

**Comment by jebbinSCL** [comment]
In a separate pull request of course.

**Pull request merged**